### ISSUE TITLE
URGENT: Dispose Method on Widgets Never Called - Massive Firebase Charges

### ISSUE THREAD

**Issue by malbrecht0792** [FlutterFlow/flutterflow-issues#3492]
### Can we access your project?

- [X] I give permission for members of the FlutterFlow team to access and test my project for the sole purpose of investigating this issue.

### Current Behavior

The dispose methods for widgets aren't being called when you navigate elsewhere in the app, so every time you navigate to a new page, it adds it to the widget tree. Consequently, each navigation triggers a rebuild of all the pages, causing all the queries and API requests to run again. This is a HUGE issue because the costs add up quickly.

For my specific scenario I have a large app with lots of pages and I'm testing it constantly going to different pages and in the background it was querying firebase and hitting my backend api which makes a google maps places API request to get location details for a location so my google cloud costs are through the roof!

The spike in costs started on July 18th (only 8 days ago) and my bill has surpassed $1000 and it's just me testing the app, I have ZERO users since I'm pre-launch.

PLEASE ESCALATE THIS ISSUE since it can impact so many users and their firebase costs!

Here is a video showing the issue:

https://www.loom.com/share/c166b8eb7b9544daaa88adb554b864d7?sid=32a94d39-cbe0-422c-aa24-53deb3b1e312

Here are my current firebase charges:

![Screenshot 2024-07-26 at 6 30 59 PM](https://github.com/user-attachments/assets/461fc785-535b-47f5-af39-0a1974f4c550)

### Expected Behavior

Widgets should dispose after navigating away. This is the behavior I see when testing a sample flutter app from the flutter repo (I tested the form_app sample app) and added the build and dispose print methods like I did in the video for the FlutterFlow app that doesn't print for dispose, and it DOES print in the dispose method on navigation for the flutter sample app

https://www.loom.com/share/9ef8e5abd487480293631d763ae1b897?sid=b4e4490e-4097-460d-ba39-1752cb68f09f

### Steps to Reproduce

Just look at the video

### Reproducible from Blank

- [X] The steps to reproduce above start from a blank project.

### Bug Report Code (Required)

https://www.loom.com/share/9ef8e5abd487480293631d763ae1b897?sid=b4e4490e-4097-460d-ba39-1752cb68f09f

### Visual documentation

I provided these above but here they are again:

https://www.loom.com/share/c166b8eb7b9544daaa88adb554b864d7?sid=32a94d39-cbe0-422c-aa24-53deb3b1e312

https://www.loom.com/share/9ef8e5abd487480293631d763ae1b897?sid=b4e4490e-4097-460d-ba39-1752cb68f09f

### Environment

```markdown
- FlutterFlow version: 4.1.78
- Platform: MacOS
- Browser name and version: Chrome 126.0.6478.182
- Operating system and version affected: Sonoma 14
```

**Comment by msusviela** [comment]
Thanks for the report! I'll send it to ENG so they can check on this

**Comment by isegal** [comment]
This is very strange. AFAIK firestore re-queries has been a known behavior for years. In fact I've also observed that having a query inside a child of a parent that has another query will rebuild and re-query the child every time the parent data changes. So a page does not even need to be navigated for the firestore clients to unnecessarily re-subscribe (churn).

The workaround for this is to enable page-based or app-based caching for firestore. Unfortunately that can bring other problems (setting custom cache keys is a must, and you need to be paying close attention to what you use for the cache keys).

I am very curious as to about your costs: it shows that it is non-firebase. These are extremely high if you are the only user of the app, so something alarming is going on. But the bill is showing that firebase and cloud functions combined are only charging 22 cents. Try drilling down into the costs from the google console cost table. It should show you a more granular breakdown of what is running up your bill.

**Comment by leighajarett** [comment]
I'm not entirely sure what "non-firebase" charges mean, it would be great if you could drill in as mentioned above!

Also +1 to the recommendation on caching. You might also consider unchecking "Disable back Navigation" in Navigate actions.

**Comment by malbrecht0792** [comment]
@leighajarett I intentionally didn't respond to the comment from @isegal because it completely misses the point of this issue, and it ran the risk of convoluting this Github issue (which it clearly has). Now that I see you're the head of product at FF, I'll respond.

This has nothing to do with caching. This has nothing to do with FF configuration. This is a bug. Why did you remove the bug label? The bug is clearly reproducible in a new project. This is very frustrating that you've changed the label without any explanation.

The "non-firebase" charges label in the firebase console are from the google maps places API. I mentioned this in the ticket and didn't want to post a screenshot of the drilled down bill in the GCP billing console for security reasons. I'm well aware of what the issue is here.

Please rewatch the videos (if you haven't already). The issue is that something with FF or possibly with go_router is causing the dispose methods of every page widget to not be called when navigating somewhere else. This means that each navigation is adding a new page to the widget tree and rebuilding ALL PAGES added. This is not the default behavior in Flutter. I show this in my second video of a regular flutter app. The dispose methods of pages are called to dispose the widgets when you navigate away. These are properly being called in the normal flutter app and the pages are disposed. But this does NOT happen in the FlutterFlow app.

To give you an idea of how this plays out in my app...I have a Home page with a feed of items. Each of those items displays a location from Google Maps. I fetch the location data using a backend endpoint I set up that uses the Google Maps Places API. The expected behavior from a functioning flutter app is that when the list re-renders or the feed item component re-renders that it calls the endpoint to fetch the location data. That would be fine with me for now. There's other optimizations I can make to this but I'm pre-launch in development and still in the GCP free tier. But that's not what's happening...

What's happening is this. Let's say for example on my Home page I have 30 feed items that each contain a query to the Google Maps Places API. When I open the app it goes to the root page (Home) and makes the 30 queries. Then when I navigate to another page in the app it makes the same 30 queries (even though we aren't on that page anymore). Every time I navigate somewhere it makes the 30 queries (for the Home page we aren't on). But it gets worse! If I navigate to Home again, it then makes 60 queries. Then if I navigate somewhere else it will make those same 60 queries. Then if I navigate to Home again it makes 90 queries. And so on and so forth. I'm testing in my app a lot and the Home page is obviously a pretty common page to navigate to. So you can see how this adds up quickly.

The reason the costs for "non-firebase" is so much higher than the other costs is because the Places API is more costly, and the free tier for firebase costs is pretty generous.

I spent hours looking into the cause of this, testing different things out and found that the issue is that the dispose method isn't being called as it should be. I reproduced this in a blank project, as shown in the video. This blank project example has nothing to do with my specific use case. It's a bare bones app example with a few pages and some buttons to navigate between them. If I'm able to reproduce the issue there, it's most likely affecting everyone.

My question is, have others reported this as well or am I the first? If this is affecting everyone then I imagine some of the power users of FF would have noticed this. Perhaps not yet though. Perhaps they haven't merged any new code since the bug started. I'm not sure. I also image some very large percentage of FF users (80, 90, 95 percent) probably just use it to build toy apps or very small apps and operate within firebase's free tier. They obviously wouldn't have noticed this issue.

If this issue is as widespread as I fear it is, then your large customers could have significant cost increases that could bankrupt them. If there's even a 1% chance of that being true, this needs to be taken extremely seriously and handled quickly.

@leighajarett please schedule a call with me to discuss this with a manager and someone from engineering. You should have my email.

Thanks

**Comment by isegal** [comment]
When pushing a page to a navigation stack, the home page is not supposed to be disposed. However, the fact that the home page frequently rebuilds when is not visible is a bit unusual.

Off the top of my head, there are several cases when the entire hierarchy rebuilds (eg when setting a global FF app state variable, or when current user document content changes).

Edit: I watched the video more closely. The behavior you are observing is related to back-navigation history.
Eg: When navigating from Page 1 -> Page 2, the reason Page 1 is not disposed is so that back navigation state can be maintained (such as scroll position if page 1 displays a list of items, form widget state if a user started filling out a form or text fields.

When going from Page 2 to page 1, if you do not use the navigate back action (aka navigator pop), the navigator will push a new page 1 to the history. This is similar to how a browser history with back navigation would work.

Here be dragons: rebuilds are meant to be cheap. Flutter's official documentation recommends NOT to make API requests inside build. [https://docs.flutter.dev/cookbook/networking/fetch-data#why-is-fetchalbum-called-in-initstate](https://docs.flutter.dev/cookbook/networking/fetch-data#why-is-fetchalbum-called-in-initstate) .

We know that FF's default behavior for Firebase queries to be invoked from within build(). If the same goes for API requests, this can be quiet a footgun as most FF users would be unaware of this.

However, FF can't simply move all queries and API requests to initState as it will cause a whole lot of other bugs (such as page information not being refreshed properly if some data dependency changes in the hierarchy).

What we need is a smarter data model that would be are when query or API parameters change and re-request only when they do.

**Comment by leighajarett** [comment]
Thanks for adding that info! Agreed with the points that are made above ^ <br><br>Happy to get on a call, I don't think I have your email though, would you be able to post it here? Otherwise you can send me an email with your availability over the next few days leigha@flutterflow.io

**Comment by malbrecht0792** [comment]
@isegal thanks for your reply. While your first comment didn't directly address my issue here, this reply is spot on. You're definitely right about this. I was wrong about this being a bug.

I did some more testing with the pure Flutter sample app and noticed that their implementation for navigation using GoRouter by default uses `go` instead of `push`, whereas FlutterFlow's default navigation configuration when you add a Navigate To Page action has the navigate back toggle turned on, which means it uses `push`.

I tested this out with my sample app and looked at the generated code. I'm sure you're aware of this but I want to write this out explicitly for clarity and in case someone else comes across this in the future. This is what happens:

With nothing toggled on...

![Screenshot 2024-07-31 at 11 47 55 AM](https://github.com/user-attachments/assets/9c283a8a-1a3c-48e5-a99f-514a6452d234)

Generates this:

```
onPressed: () async {
  context.goNamed('page2');
},
```

Toggling on just the "Navigate Back" toggle...

![Screenshot 2024-07-31 at 11 47 46 AM](https://github.com/user-attachments/assets/80a6ba39-654d-40bd-9295-631140b17ce7)

Generates this:

```
onPressed: () async {
    context.pushNamed('page2');
},
```

Toggling on the "Navigate Back" toggle AND the "Replace Route" toggle...

![Screenshot 2024-07-31 at 11 47 59 AM](https://github.com/user-attachments/assets/34df49b4-f526-4b4f-a877-f1e1ec3adff4)

Generates this:

```
onPressed: () async {
  if (Navigator.of(context).canPop()) {
    context.pop();
  }
  context.pushNamed('page2');
},
```

The default configuration when adding a Navigate To Page action toggles on the Allow Back Navigation setting, which means it's implementing `pushNamed`. This seems dangerous to me for the specific reason I got into this situation and have a massive GCP bill now.

I think by default Allow Back Navigation should be toggled OFF and when you go to toggle it on it should give display a warning explaining that this will keep adding pages to the navigation stack and updates will trigger a rebuild of each individual instance of the page's build method (re-querying any firestore queries or API queries defined in the widget tree).

It also seems from the FlutterFlow documentation that adding data queries (firestore, api, etc) inside the widget tree (not in initState) is the recommended way of implementing queries. I actually changed the implementation of my app to be more like this because I thought it was the "right" way to do things, not knowing about the `pushNamed` navigation re-adding the pages to the navigation stack and rebuilding all the time. Before, I had implemented it in the "On Page Load" (initState) section and was saving to page/component state, but then changed it to being in the widget tree.

I feel like this should be explained more clearly in the FlutterFlow docs, perhaps in a warning section, since it can possibly be expensive depending on what data you're fetching and how many pages use the Allow Back Navigation setting.

**Comment by malbrecht0792** [comment]
@leighajarett regarding this GitHub issue, it's not a bug. But I still have a GCP bill of over $1000 stemming from the misconfiguration of my app (not realizing that it's re-running all my data queries). Is it possible for me to get this waived somehow?

It's almost all due to the Google Maps Places API which luckily doesn't really cost Google much since it's just a data API request (unlike tying up cloud resources with a Cloud Function for example). I know that FlutterFlow has a close relationship with Google and I'm wondering if there's any way you're able to let them know this was caused by a misconfiguration in FlutterFlow.

Like I mentioned before, my app isn't even launched yet and it's just me testing out my app and now I have a bill of over $1000. If there's any way you can help out here, I would GREATLY appreciate it.

Thanks

**Comment by isegal** [comment]
> It also seems from the FlutterFlow documentation that adding data queries (firestore, api, etc) inside the widget tree (not in initState) is the recommended way of implementing queries.

An interesting question is why this recommendation is contradicting the official Flutter documentation.

If I were to guess, this is likely because the alternative will also lead to surprising behavior: If you have a Parent with an auto-refresh (aka listen) query A, and a Child with a query B, if B depends on data from A placing B in initState will make it stale when A's data refreshes.

FlutterFlow provides query caching, but this has some problems of its own:
1. There is a hard-coded limit of number of cache queries (see `request_manager.dart`):
```
class FutureRequestManager<T> {
  FutureRequestManager([this.cacheLimit = 10]);
```
2. More importantly, the default cache request key is hard-coded, so if you have the same cache across different pages or elements, if the query filters differ then you will get strange bugs with stale data:
```
String _requestKey(String? key) =>
    key == null || key.isEmpty ? '__DEFAULT_KEY__' : key;
```

What would be really nice from FlutterFlow is a better default cache keying where it depends on the firestore query(request) object itself rather than a hardcoded string.

Such query key generator could look something like:


Where defaultKey() would create a unique hash based on the query itself (including all the filters and parameters), and could be defined as:

```
String defaultKey(Query query) {
   return query.hashCode.toString();
}
```

Note: If hashCode in Fitestore's Query implementation does not represent the query payload (eg filters and all the settings/params), then an altenative would be to simply:

```
String defaultKey(Query query) {
   return query.toString();
}
```

The default query key generation would then be based on the query filter, including all the `where()` statements, `limit()`, etc. 

```
String _requestKey(String? key, Query query) =>
    key == null || key.isEmpty ? defaultKey(query) : key;
```

Going back to the original example, this would prevent stale cache, because if query B has any filters that depend on query A, a data refresh on A would cause the filter cache key to change, and the data dependency would properly re-query. As long as the filters/values do not change, rebuilds would return the existing cached query.

Note: the problem with the default cache key only exists if you have are using the same cache across different pages, or across different elements within a page and your filters differ.

In summary - there are unfortunately many footguns that a developer needs to be aware of. Some of them can be solved (and should) by FlutterFlow team.

**Comment by malbrecht0792** [comment]
@isegal Yeah I'm not sure why the FF documentation contradicts the flutter docs on this. Although, I'm also not sure FF implementing this sort of caching is necessary for this use case either, since you can also use a snapshot listener to achieve this. For example, listening for new items in a feed and then updating app state:

```
import 'package:firebase_auth/firebase_auth.dart';

Future listenForUserFeedChanges() async {
  final currentUser = FirebaseAuth.instance.currentUser;
  if (currentUser == null) {
    return;
  }
  FirebaseFirestore.instance
      .collection('users')
      .doc(currentUser.uid)
      .collection('user_feed')
      .snapshots()
      .listen((snapshot) {
    if (snapshot.docChanges.isNotEmpty) {
      print('Changes detected in user feed.');
      FFAppState().hasNewFeedItems = true;
    }
  });
}
```

I also think that stale data is a much better _problem_ to have then re-querying on every build, especially since one is obvious and the other isn't unless you're looking at the network requests which many less experienced FF developers probably might not be.

I've definitely learned my lesson with this though and appreciate your input here. I'll be at FFDC in September so maybe I'll see you there!
